L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,471 @@
+L46: C++: New TLS Credentials API
+----
+* Author(s): gtcooke94, ZhenLian
+* Approver: 
+* Status: Draft
+* Implemented in:
+* Last updated: May 16, 2024
+
+## Abstract
+
+This proposal aims to provide a new TLS credentials API with the following features:
+
+1. Credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without restarting the process.
+1. Customizable verification of the peer certificate: perform custom validation checks on the peer's certificate chain after it has been verified to chain up to a root of trust, or fully customize path building and cryptographic verification.
+1. TLS version configuration: optionally configure the minimum and maximum of the TLS versions to be negotiated during the handshake.
+1. Support for certificate revocation via Certificate Revocation Lists (CRLs).
+
+## Background
+
+The current TLS implementation (the SSL credentials API) in gRPC C core has some restrictions:
+1. Lack of the credential reloading support. Once an application starts, it can never reload its private key, certificate chain conveying the application's identity, and root certificates without shutting down.
+1. Inflexibility of how the peer certificate chain is verified. There is no API to add additional custom verification checks.
+1. Inability to choose the maximum/minimum TLS version to be negotiated during the TLS handshake.
+1. No support for certificate revocation.
+
+We propose an API that supports the above features and meets the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones do not fit their needs.
+1. The credential reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection.
+1. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API.
+
+### Related Proposals:
+
+1. [L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98) - this was subsumed into this proposal.
+1. [A69: Crl Providers](https://github.com/grpc/proposal/pull/382)
+
+## Proposal
+Until recently, C-Core needed to be written in C89, then we had C++ wrappers of those APIs in `grpcpp`. Requirements have changed such that we can use C++ directly in Core. The rest of this proposal is written as C++ that will be implemented in core, and simple aliases to everything will be written in the `grpcpp`.
```

Comment:
optional suggested edit: s/C-Core/the C-core boundary APIs

Reply:
Done